### PR TITLE
Feat/androidkeystore

### PR DESCRIPTION
**Pull request by SeryogaBrigada**

### CODE REVIEW

File: src/libaktualizr/crypto/keymanager.cc [hunk 1]
```diff
@@ -209,11 +224,21 @@ Json::Value KeyManager::signTuf(const Json::Value &in_data) const {
     crypto_engine = (*p11_)->getEngine();
     private_key = config_.p11.uptane_key_id;
   }
+
+  std::string b64sig;
   if (config_.uptane_key_source == CryptoSource::kFile) {
     backend_->loadPrimaryPrivate(&private_key);
+      b64sig = Utils::toBase64(
```
Comment (OYTIS) [inline src/libaktualizr/crypto/keymanager.cc]: Some formatting problems here.

File: src/libaktualizr/crypto/keymanager.cc [hunk 2]
```diff
@@ -230,15 +255,29 @@ std::string KeyManager::generateUptaneKeyPair() {
   std::string primary_public;
 
   if (config_.uptane_key_source == CryptoSource::kFile) {
-    std::string primary_private;
-    if (!backend_->loadPrimaryKeys(&primary_public, &primary_private)) {
-      if (Crypto::generateKeyPair(config_.uptane_key_type, &primary_public, &primary_private)) {
-        backend_->storePrimaryKeys(primary_public, primary_private);
+      std::string primary_private;
```
Comment (OYTIS) [inline src/libaktualizr/crypto/keymanager.cc]: And here.

File: src/libaktualizr/uptane/fetcher.cc [hunk 3]
```diff
@@ -23,8 +23,8 @@ bool Fetcher::fetchRole(std::string* result, int64_t maxsize, RepositoryType rep
 static size_t DownloadHandler(char* contents, size_t size, size_t nmemb, void* userp) {
   assert(userp);
   auto* ds = static_cast<Uptane::DownloadMetaStruct*>(userp);
-  uint64_t downloaded = size * nmemb;
-  uint64_t expected = ds->target.length();
+  size_t downloaded = size * nmemb;
```
Comment (OYTIS) [inline src/libaktualizr/uptane/fetcher.cc]: Why changing the type? I hoped we still could be able to download images larger than 4Gb on 32 bit devices. There are no tests for that right now, but let's not break that deliberately.

File: src/libaktualizr/uptane/ipsecondarydiscovery.cc [hunk 4]
```diff
@@ -93,7 +93,7 @@ std::vector<Uptane::SecondaryConfig> IpSecondaryDiscovery::waitDevices() {
     if (left_seconds <= 0) {
       break;
     }
-    tv.tv_sec = left_seconds;
+    tv.tv_sec = static_cast<long>(left_seconds);
```
Comment (OYTIS) [inline src/libaktualizr/uptane/ipsecondarydiscovery.cc]: `static_cast<time_t>` ?

File: src/libaktualizr/uptane/fetcher.cc (hunk 3 above)
Comment (SeryogaBrigada) [reply src/libaktualizr/uptane/fetcher.cc]: OK, I'll remove the last commit in the branch

**Comment by OYTIS** [comment]
I like the overall direction. Any idea if/how we could tests that on HiKey?

**Comment by SeryogaBrigada** [comment]
After merging with the feat/android-support branch, the changes will take effect.

File: src/libaktualizr/crypto/keymanager.cc (hunk 1 above)
Comment (SeryogaBrigada) [reply src/libaktualizr/crypto/keymanager.cc]: I'll check it

**Comment by SeryogaBrigada** [comment]
Formatting problems fixed, "Fix build for 32 bit" commit removed.

**Comment by SeryogaBrigada** [comment]
How we can test it? Since the functions runs at runtime, we can monitor them with a debug message.
For example, we can add several lines of code:
```
__android_log_write(ANDROID_LOG_DEBUG, "Test getPublicKey()", AndroidKeyStore::instance().getPublicKey().c_str());
__android_log_write(ANDROID_LOG_DEBUG, "Test generateKeyPair()", AndroidKeyStore::instance().generateKeyPair().c_str());
std::string data = AndroidKeyStore::instance().signData("TestData");
__android_log_write(ANDROID_LOG_DEBUG, "Test signData()", data.c_str());
bool res = AndroidKeyStore::instance().verifyData("TestData", data);
__android_log_write(ANDROID_LOG_DEBUG, "Test verifyData()", res ? "Yes" : "No");
```

to the aktualizrJNIwrapper.cpp file to receive debug messages.

File: src/libaktualizr/crypto/keymanager.cc [hunk 5]
```diff
@@ -209,11 +224,21 @@ Json::Value KeyManager::signTuf(const Json::Value &in_data) const {
...
+            Crypto::Sign(config_.uptane_key_type, crypto_engine, private_key, Json::FastWriter().write(in_data)));
+  }
+  if (config_.uptane_key_source == CryptoSource::kAndroid) {
+#if defined(ANDROID)
+    b64sig = AndroidKeyStore::instance().signData(Json::FastWriter().write(in_data));
+#else
+    throw std::runtime_error("Aktualizr was built without Android support");
+#endif
   }
-  std::string b64sig = Utils::toBase64(
```
Comment (lbonn) [inline src/libaktualizr/crypto/keymanager.cc]: Moving this line broke P11 support

Comment (lbonn) [reply src/libaktualizr/crypto/keymanager.cc]: This "patch" should fix the issue:

```
diff --git a/src/libaktualizr/crypto/keymanager.cc b/src/libaktualizr/crypto/keymanager.cc
index 722cf538..8dca70c9 100644
--- a/src/libaktualizr/crypto/keymanager.cc
+++ b/src/libaktualizr/crypto/keymanager.cc
@@ -226,17 +226,20 @@ Json::Value KeyManager::signTuf(const Json::Value &in_data) const {
   }
 
   std::string b64sig;
-  if (config_.uptane_key_source == CryptoSource::kFile) {
-    backend_->loadPrimaryPrivate(&private_key);
-    b64sig = Utils::toBase64(
-            Crypto::Sign(config_.uptane_key_type, crypto_engine, private_key, Json::FastWriter().write(in_data)));
-  }
   if (config_.uptane_key_source == CryptoSource::kAndroid) {
 #if defined(ANDROID)
     b64sig = AndroidKeyStore::instance().signData(Json::FastWriter().write(in_data));
 #else
     throw std::runtime_error("Aktualizr was built without Android support");
 #endif
+  } else {
+    if (config_.uptane_key_source == CryptoSource::kFile) {
+      backend_->loadPrimaryPrivate(&private_key);
+    }
+    // no need to load key in p11 case
+
+    b64sig = Utils::toBase64(
+            Crypto::Sign(config_.uptane_key_type, crypto_engine, private_key, Json::FastWriter().write(in_data)));
   }
 
   Json::Value signature;
```

Comment (SeryogaBrigada) [reply src/libaktualizr/crypto/keymanager.cc]: OK, I'll check it.

**Comment by SeryogaBrigada** [comment]
Fixed and rebased over the latest master

**Comment by lbonn** [comment]
Thanks! There are still formatting issues. If you amend your commit after `make format`, it should be good to go.

**Comment by SeryogaBrigada** [comment]
Formatting done.

**Comment by lbonn** [comment]
Unfortunately, it seems to differ again with what the CI expects. It could be a problem with your clang-format version (we expect clang-format-6.0). You will probably need to re-run cmake from scratch after installing it.

If needed, you can also use one the included docker image:

```
./scripts/run_docker_test.sh docker/Dockerfile.debian.testing -- sh -c 'mkdir build-format; cd build-format; cmake ..; make -j8 format'
```

**Comment by SeryogaBrigada** [comment]
Fixed.

**Pull request merged**